### PR TITLE
#134:  Improve support for HTTP/2

### PR DESCRIPTION
**Pull request by jowerner**

### CODE REVIEW

File: src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+        // basic settings and components
+        httpClientBuilder.authenticator(authenticator);
+        // httpClientBuilder.cache(null);
+        httpClientBuilder.connectionPool(connectionPool);
+        // httpClientBuilder.connectionSpecs(createConnectionSpecs(webClientOptions));
+        httpClientBuilder.cookieJar(new CookieJarImpl(webClient.getCookieManager()));
+        httpClientBuilder.dns(dns);
+        httpClientBuilder.followRedirects(false);
+        httpClientBuilder.followSslRedirects(false);
```
Comment (h-arlt) [inline src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java:129]: Maybe we should add a comment why we disable automatic following of redirects.

File: src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+        // timeouts
+        int timeout = webRequest.getTimeout();
+        if (timeout < 0)
+        {
+            timeout = webClientOptions.getTimeout();
+        }
+
+        final Duration timeoutMS = Duration.ofMillis(timeout);
+        // httpClientBuilder.callTimeout(timeoutMS);
```
Comment (h-arlt) [inline src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java:152]: Remove the configuration statements that have been commented out?

File: src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+        else if (pairWithFile.getFileName() == null)
+        {
+            filename = file.getName();
+        }
+        else
+        {
+            filename = pairWithFile.getFileName();
+        }
+
+        // now do the right thing
```
Comment (h-arlt) [inline src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java:274]:
```suggestion
        // add form data part in the right way (depending on what was given as input)
```

File: src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+import com.gargoylesoftware.htmlunit.util.NameValuePair;
+import com.gargoylesoftware.htmlunit.util.UrlUtils;
+import com.xceptance.xlt.api.util.XltException;
+import com.xceptance.xlt.engine.util.TimerUtils;
+
+/**
+ * Super class for alternative {@link WebConnection} implementations. Provides some common functionality that we would
+ * otherwise have to implement again and again.
+ */
+public abstract class AbstractWebConnection<HttpClient, Request, Response> implements WebConnection
```
[the remark this answers is not in the document]
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java]: Roger, done.

File: src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    protected abstract Request createRequestWithMultiPartBody(final URI uri, final WebRequest webRequest);
+
+    /**
+     * Executes the implementation-specific Request instance using the given HttpClient object and returns a
+     * corresponding Response object.
+     */
+    protected abstract Response executeRequest(HttpClient httpClient, final Request request) throws IOException;
+
+    /**
+     * Creates a {@link WebResponse} instance from the implementation-specific Response and links it to the given
```
[the remark this answers is not in the document]
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java]: Done.

File: src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+                                                           final @Nullable Charset charset);
+
+    /**
+     * Creates an implementation-specific multi-part Request instance that resembles the given {@link WebRequest}.
+     */
+    protected abstract Request createRequestWithMultiPartBody(final URI uri, final WebRequest webRequest);
+
+    /**
+     * Executes the implementation-specific Request instance using the given HttpClient object and returns a
+     * corresponding Response object.
```
[the remark this answers is not in the document]
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java]: Done.

File: src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+    /**
+     * Creates an implementation-specific Request instance that resembles the given {@link WebRequest} and populates it
+     * with the given string body.
+     */
+    protected abstract Request createRequestWithStringBody(final URI uri, WebRequest webRequest, final String body, final String mimeType,
+                                                           final @Nullable Charset charset);
+
+    /**
+     * Creates an implementation-specific multi-part Request instance that resembles the given {@link WebRequest}.
```
[the remark this answers is not in the document]
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java]: Done.

File: src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        return request;
+    }
+
+    /**
+     * Creates an implementation-specific Request instance that resembles the given {@link WebRequest} and has no body.
+     */
+    protected abstract Request createRequestWithoutBody(final URI uri, final WebRequest webRequest);
+
+    /**
+     * Creates an implementation-specific Request instance that resembles the given {@link WebRequest} and populates it
```
[the remark this answers is not in the document]
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java]: Done.

File: src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+    protected abstract WebResponse makeWebResponse(final Response response, final WebRequest webRequest, final long loadTime)
+        throws IOException;
+
+    /**
+     * Adds some standard headers to the web request.
+     * 
+     * @param webRequest
+     *            the web request
+     */
+    private void finalizeRequestHeaders(final WebRequest webRequest)
```
[the remark this answers is not in the document]
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/engine/htmlunit/AbstractWebConnection.java]: Yeah, may come in handy. Done.

File: src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/AuthenticatorImpl.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        // check if a previous (proxy) authentication attempt failed
+        if (statusCode == 401 && response.request().header(HttpRequestHeaders.AUTHORIZATION) != null
+            || statusCode == 407 && response.request().header(HttpRequestHeaders.PROXY_AUTHORIZATION) != null)
+        {
+            // give up, we've already failed to authenticate
+            return null;
+        }
+
+        // whether to authenticate with a proxy or an origin server
+        final boolean isProxyAuth = statusCode == 407;
```
[the remark this answers is not in the document]
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/AuthenticatorImpl.java]: Done.

File: src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java (hunk 3 above)
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java]: Done.

File: src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java (hunk 1 above)
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java]: Done.

File: src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java (hunk 2 above)
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java]: Done.

**Pull request merged**